File: backend/models/procedure.py
from sqlalchemy import JSON, Boolean, Float, ForeignKey, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from backend.db.base import Base


class Procedure(Base):
    __tablename__ = "procedures"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    slug: Mapped[str] = mapped_column(String(100), unique=True, nullable=False, index=True)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    category: Mapped[str] = mapped_column(String(100), nullable=False)
    description: Mapped[str] = mapped_column(Text, nullable=False)
    ente_competente: Mapped[str] = mapped_column(String(200), nullable=False)
    tags: Mapped[list] = mapped_column(JSON, nullable=False, default=list)
    tempo_stimato_giorni: Mapped[int | None] = mapped_column(Integer, nullable=True)
    costo_stimato_eur: Mapped[float | None] = mapped_column(Float, nullable=True)
    active: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)

    steps: Mapped[list["Step"]] = relationship(
        "Step", back_populates="procedure", order_by="Step.order", cascade="all, delete-orphan"
    )


class Step(Base):
    __tablename__ = "steps"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    procedure_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("procedures.id", ondelete="CASCADE"), nullable=False
    )
    order: Mapped[int] = mapped_column(Integer, nullable=False)
    title: Mapped[str] = mapped_column(String(200), nullable=False)
    description: Mapped[str] = mapped_column(Text, nullable=False)
    ufficio: Mapped[str | None] = mapped_column(String(200), nullable=True)
    costo_stimato_eur: Mapped[float | None] = mapped_column(Float, nullable=True)
    tempo_stimato_giorni: Mapped[int | None] = mapped_column(Integer, nullable=True)

    procedure: Mapped["Procedure"] = relationship("Procedure", back_populates="steps")
    documents: Mapped[list["Document"]] = relationship(
        "Document", back_populates="step", order_by="Document.id", cascade="all, delete-orphan"
    )


class Document(Base):
    __tablename__ = "documents"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    step_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("steps.id", ondelete="CASCADE"), nullable=False
    )
    name: Mapped[str] = mapped_column(String(300), nullable=False)
    required: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)

    step: Mapped["Step"] = relationship("Step", back_populates="documents")

File: mobile/lib/models/procedure.dart
class ProcedureDocument {
  final String name;
  final bool required;
  final String? notes;

  const ProcedureDocument({
    required this.name,
    required this.required,
    this.notes,
  });

  factory ProcedureDocument.fromJson(Map<String, dynamic> json) =>
      ProcedureDocument(
        name: json['name'] as String,
        required: json['required'] as bool,
        notes: json['notes'] as String?,
      );
}

class ProcedureStep {
  final int order;
  final String title;
  final String description;
  final String? ufficio;
  final double? costoStimatoEur;
  final int? tempoStimatoGiorni;
  final List<ProcedureDocument> documents;

  const ProcedureStep({
    required this.order,
    required this.title,
    required this.description,
    this.ufficio,
    this.costoStimatoEur,
    this.tempoStimatoGiorni,
    required this.documents,
  });

  factory ProcedureStep.fromJson(Map<String, dynamic> json) => ProcedureStep(
        order: json['order'] as int,
        title: json['title'] as String,
        description: json['description'] as String,
        ufficio: json['ufficio'] as String?,
        costoStimatoEur: (json['costo_stimato_eur'] as num?)?.toDouble(),
        tempoStimatoGiorni: json['tempo_stimato_giorni'] as int?,
        documents: (json['documents'] as List<dynamic>)
            .map((d) => ProcedureDocument.fromJson(d as Map<String, dynamic>))
            .toList(),
      );
}

class Procedure {
  final String slug;
  final String name;
  final String category;
  final String description;
  final String enteCompetente;
  final List<String> tags;
  final int? tempoStimatoGiorni;
  final double? costoStimatoEur;
  final List<ProcedureStep> steps;

  const Procedure({
    required this.slug,
    required this.name,
    required this.category,
    required this.description,
    required this.enteCompetente,
    required this.tags,
    this.tempoStimatoGiorni,
    this.costoStimatoEur,
    this.steps = const [],
  });

  factory Procedure.fromJson(Map<String, dynamic> json) => Procedure(
        slug: json['slug'] as String,
        name: json['name'] as String,
        category: json['category'] as String,
        description: json['description'] as String,
        enteCompetente: json['ente_competente'] as String,
        tags: List<String>.from(json['tags'] as List<dynamic>),
        tempoStimatoGiorni: json['tempo_stimato_giorni'] as int?,
        costoStimatoEur: (json['costo_stimato_eur'] as num?)?.toDouble(),
        steps: json.containsKey('steps')
            ? (json['steps'] as List<dynamic>)
                .map((s) => ProcedureStep.fromJson(s as Map<String, dynamic>))
                .toList()
            : const [],
      );

  Map<String, dynamic> toJson() => {
        'slug': slug,
        'name': name,
        'category': category,
        'description': description,
        'ente_competente': enteCompetente,
        'tags': tags,
        'tempo_stimato_giorni': tempoStimatoGiorni,
        'costo_stimato_eur': costoStimatoEur,
      };
}
